### PR TITLE
[KYUUBI #1300] Detecting critical errors

### PR DESCRIPTION
**Pull request by yaooqinn**
### _Why are the changes needed?_

For critical errors at engine side, it is not handled properly. For example,when engine oom
- server may not be able to get operation statuses because of no response of engine side. In this case, client only get a ambiguous `read timeout` as a final cause.
- the oom hook of engine side might directly crash the engine, when the server is still trying to get operation status

In this PR,
- a config for retry to make the operation status updating process more robust
- make the engine oom hook only de register itself to make it able to recover for some transient errors 

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by yaooqinn** [comment]
cc @turboFei @pan3793 @zhouyifan279

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala [hunk 1]
```diff
@@ -83,18 +83,39 @@ class ExecuteStatement(
...
+      } catch {
+        case e: TException if currentAttempts >= maxStatusPollOnFailure =>
+          error(s"Failed to get ${session.user}'s query[$getHandle] status after" +
+            s" $maxStatusPollOnFailure times, aborting", e)
+          throw e
+        case e: TException =>
+          currentAttempts += 1
+          warn(s"Failed to get ${session.user}'s query[$getHandle] status" +
+            s" ($currentAttempts / $maxStatusPollOnFailure)", e)
+          Thread.sleep(500)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala:102]: Configurable?

**Review by pan3793 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala (hunk 1 above)
Comment (yaooqinn) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/operation/ExecuteStatement.scala]: Look fine to me to be hardcoded. As users already have operation.status.polling.timeout

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendService.scala [hunk 2]
```diff
@@ -147,6 +147,15 @@ class SparkThriftBinaryFrontendService(
       s"${serverAddr.getHostAddress}:$portNum"
     }
   }
+
+  // When a OOM occurs, here we de-register the engine by stop its discoveryService.
+  // Then the current engine will not be connected by new client anymore but keep the existing ones
+  // alive. In this case we can reduce the engine's overhead and make it possible recover from that.
+  // We shall not tear down the whole engine by serverable.stop to make the engine unreachable for
+  // the exising clients which are still get statuses and report the end-users.
```
Comment (zhouyifan279) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/SparkThriftBinaryFrontendService.scala:155]: 1. `exising` -> `existing`
2. `report the end-users` -> `report to the end-users`

**Comment by pan3793** [comment]
Thanks, merging to master

**Pull request closed without merging**